flow/FastAlloc.h

```diff
@@ -22,248 +22,127 @@
-#define INSTRUMENT_RELEASE(name) (allocInstr[(name)].dealloc())
-
-// extern std::map<uint32_t, uint64_t> stackAllocations;
-
-// maps from an address to the hash of the backtrace and the size of the alloction
-extern std::unordered_map<int64_t, std::pair<uint32_t, size_t>> memSample;
-
-struct BackTraceAccount {
-	double count;
-	size_t sampleCount;
-	size_t totalSize;
-	std::vector<void*>* backTrace;
-};
-// maps from a hash of a backtrace to a backtrace and the total size of data currently allocated from this stack
-extern std::unordered_map<uint32_t, BackTraceAccount> backTraceLookup;
-
-extern ThreadSpinLock memLock;
-extern thread_local bool memSample_entered;
-extern const size_t SAMPLE_BYTES;
-
-#else
-#define INSTRUMENT_ALLOCATE(name)
-#define INSTRUMENT_RELEASE(name)
-#endif
-
-#if defined(ALLOC_INSTRUMENTATION) || defined(ALLOC_INSTRUMENTATION_STDOUT)
-void recordAllocation(void* ptr, size_t size);
-void recordDeallocation(void* ptr);
-#endif
-
-template <int Size>
-class FastAllocator {
-public:
-	[[nodiscard]] static void* allocate();
-	static void release(void* ptr);
-	static void check(void* ptr, bool alloc);
-
-	static long long getTotalMemory();
-	static long long getApproximateMemoryUnused();
-	static long long getActiveThreads();
-
-	static void releaseThreadMagazines();
-
-#ifdef ALLOC_INSTRUMENTATION
-	static volatile int32_t pageCount;
-#endif
-
-	FastAllocator() = delete;
-
-private:
-#ifdef VALGRIND
-	static unsigned long vLock;
-#endif
-
-	static const int magazine_size = (128 << 10) / Size;
-	static const int PSize = Size / sizeof(void*);
-	struct GlobalData;
-	struct ThreadData {
-		void* freelist;
-		int count; // there are count items on freelist
-		void* alternate; // alternate is either a full magazine, or an empty one
-	};
-	static thread_local ThreadData threadData;
-	static thread_local bool threadInitialized;
-	static GlobalData* globalData() noexcept {
-#ifdef VALGRIND
-		ANNOTATE_RWLOCK_ACQUIRED(vLock, 1);
-#endif
-		static GlobalData* data = new GlobalData(); // This is thread-safe as of c++11 (VS 2015, gcc 4.8, clang 3.3)
-
-#ifdef VALGRIND
-		ANNOTATE_RWLOCK_RELEASED(vLock, 1);
-#endif
-
-		return data;
-	}
-	static void* freelist;
-
-	static void initThread();
-	static void getMagazine();
-	static void releaseMagazine(void*);
-};
-
-extern std::atomic<int64_t> g_hugeArenaMemory;
-void hugeArenaSample(int size);
-void releaseAllThreadMagazines();
-int64_t getTotalUnusedAllocatedMemory();
-void setFastAllocatorThreadInitFunction(
-    void (*)()); // The given function will be called at least once in each thread that allocates from a FastAllocator.
-                 // Currently just one such function is tracked.
+#include "flow/Error.h"
+#include <cstdint>
+#include <cstddef>
 
 inline constexpr int nextFastAllocatedSize(int x) {
-	assert(x > 0 && x <= 8192);
+	// Based on http://jemalloc.net/jemalloc.3.html#size_classes
+	ASSERT(0 < x && x <= 8192);
+	if (x <= 8)
```

Comment:
Nit: would it make sense to replace this code with a look-up table or something? I guess a `std::map::lower_bound` would be readable but slow.

Reply:
I _think_ I did this because it needs to be constexpr. I can try using std::lower_bound with an array and see if that works

Reply:
Oh I missed the `constexpr` in that case this makes sense